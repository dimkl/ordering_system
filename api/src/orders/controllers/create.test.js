/**
 * @integration-test true
 * @data-factory true
 */
const Order = require("../models/order");

describe("POST /orders", () => {
  beforeAll(() => require('../../shared/setupModels')());
  beforeEach(() => Order.knex().raw('truncate orders, order_items, customers, users, products, time_slots, slots cascade;'));
  afterAll(() => Order.knex().destroy());

  it("creates and returns a order", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request.post("/orders")
      .send({ customer_id: timeSlot.customer.uuid, time_slot_id: timeSlot.uuid })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot({
      id: expect.any(Number),
      uuid: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String),
      customer_id: expect.any(String),
      time_slot_id: expect.any(String)
    });
    expect(response.body.customer_id).toEqual(timeSlot.customer.uuid);
  });

  it("creates and returns a order using internal customer_id", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request.post("/orders")
      .send({ customer_id: timeSlot.customer.id, time_slot_id: timeSlot.id })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot({
      id: expect.any(Number),
      uuid: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String),
      customer_id: expect.any(String),
      time_slot_id: expect.any(String)
    });
    expect(response.body.customer_id).toEqual(timeSlot.customer.uuid);
  });

  it("throws validation error for required properties", async () => {
    const response = await request.post("/orders")
      .send({})
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it("throws validation error for not existing customer_id", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request.post("/orders")
      .send({ customer_id: timeSlot.customer.uuid + '1', time_slot_id: timeSlot.uuid })
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it("throws validation error for not existing time_slot_id", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request.post("/orders")
      .send({ customer_id: timeSlot.customer.uuid, time_slot_id: timeSlot.uuid  + '1' })
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it("throws validation error with additional properties", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request.post("/orders")
      .send({ customer_id: timeSlot.customer.uuid, time_slot_id: timeSlot.uuid, created_at: Date.now() })
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });
});