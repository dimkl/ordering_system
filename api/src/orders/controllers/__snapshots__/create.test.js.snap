// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /orders creates and returns a order 1`] = `
Object {
  "created_at": Any<String>,
  "customer_id": Any<String>,
  "id": Any<Number>,
  "state": "draft",
  "time_slot_id": Any<String>,
  "updated_at": Any<String>,
  "uuid": Any<String>,
}
`;

exports[`POST /orders creates and returns a order using internal customer_id 1`] = `
Object {
  "created_at": Any<String>,
  "customer_id": Any<String>,
  "id": Any<Number>,
  "state": "draft",
  "time_slot_id": Any<String>,
  "updated_at": Any<String>,
  "uuid": Any<String>,
}
`;

exports[`POST /orders throws validation error for not existing customer_id 1`] = `
Array [
  Object {
    "instancePath": "/customer_id",
    "keyword": "type",
    "message": "must be integer",
    "params": Object {
      "type": "integer",
    },
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf/0/type",
  },
  Object {
    "instancePath": "/customer_id",
    "keyword": "format",
    "message": "must match format \\"uuid\\"",
    "params": Object {
      "format": "uuid",
    },
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf/1/format",
  },
  Object {
    "instancePath": "/customer_id",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": Object {},
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf",
  },
]
`;

exports[`POST /orders throws validation error for not existing time_slot_id 1`] = `
Array [
  Object {
    "instancePath": "/time_slot_id",
    "keyword": "type",
    "message": "must be integer",
    "params": Object {
      "type": "integer",
    },
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf/0/type",
  },
  Object {
    "instancePath": "/time_slot_id",
    "keyword": "format",
    "message": "must match format \\"uuid\\"",
    "params": Object {
      "format": "uuid",
    },
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf/1/format",
  },
  Object {
    "instancePath": "/time_slot_id",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": Object {},
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf",
  },
]
`;

exports[`POST /orders throws validation error for required properties 1`] = `
Array [
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'customer_id'",
    "params": Object {
      "missingProperty": "customer_id",
    },
    "schemaPath": "#/required",
  },
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'time_slot_id'",
    "params": Object {
      "missingProperty": "time_slot_id",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`POST /orders throws validation error with additional properties 1`] = `
Array [
  Object {
    "instancePath": "",
    "keyword": "additionalProperties",
    "message": "must NOT have additional properties",
    "params": Object {
      "additionalProperty": "created_at",
    },
    "schemaPath": "#/additionalProperties",
  },
]
`;
